src/integrals.jl

```diff
@@ -41,44 +259,98 @@ function simpson_weight_array(n::Int; T=Float64)
-    end
+    K = combine_kernels(ProbeA, ProbeB, bkgq, R)
 
     #Integration in χ is performed using the Simpson quadrature rule
-    Δχ = ((last(χ)-first(χ))/(nχ-1))
+    Δχ = ((last(bkgq.χz_array)-first(bkgq.χz_array))/(nχ-1))
     w_χ = simpson_weight_array(nχ)
 
     #Integration in R is performed using the Clenshaw-Curtis quadrature rule
-    #CC_obj = FastTransforms.chebyshevmoments1(Float64, 2*nR+1)
-    #w_R = FastTransforms.clenshawcurtisweights(CC_obj)
-    w_R = get_clencurt_weights(-1, 1, 2*nR+1 )
-    w_R = w_R[nR+2:end]
-    w_R[1]/=2 #TODO: investigate if there are better solutions, this is not the analytic solution.
+    w_R = get_clencurt_weights_R_integration(2*nR+1)
+
+    ell_prefactor = get_ell_prefactor(ProbeA, ProbeB, ℓ_list)
+
+    @tullio Cℓ[l,i,j] := ell_prefactor[l]*bkgq.χz_array[n]*K[i,j,n,m]*w[l,n,m]*w_χ[n]*w_R[m]*Δχ
+
+    return Cℓ 
+end
 
-    @tullio Cℓ[l,i,j] := χ[n]*K[i,j,n,m]*w[l,n,m]*w_χ[n]*w_R[m]*Δχ
+"""
+    compute_Cℓ(w::AbstractArray{T, 3}, 
+               K::AbstractArray{T, 4}, 
+               bkgq::BackgroundQuantities, 
+               weights_χ::AbstractArray{T, 1}, 
+               weights_R::AbstractArray{T, 1}, 
+               ell_prefactor::AbstractArray{T,1}) where T
+
+Computes the angular power spectrum `Cℓ` by performing integrals over `χ` and `R` for a given set of inputs.
+
+# Arguments
+- `w::AbstractArray{T, 3}`: A 3D array representing the projected matter densities, containing the inner integrals over `k`. The array dimensions are (ℓ, χ, R).
+- `K::AbstractArray{T, 4}`: A 4D array representing the kernel product `K[i, j, χ, R]` for the given cosmological probes. It combines the effects of the probes and tomographic bin combinations.
+- `bkgq::BackgroundQuantities`: Contains background information, including `χz_array` for distances in the cosmology.
+- `weights_χ::AbstractArray{T, 1}`: A 1D array of weights for Simpson quadrature integration over `χ`.
+- `weights_R::AbstractArray{T, 1}`: A 1D array of weights for Clenshaw-Curtis quadrature integration over `R`.
+- `ell_prefactor::AbstractArray{T,1}`: A 1D array of prefactors dependent on angular multipole values `ℓ`, combining with other components to form the power spectrum.
 
-    return Cℓ
+# Returns
+- A 3D array `Cℓ` with dimensions (ℓ, i, j), where `i` and `j` represent the tomographic bins. The array contains the computed angular power spectrum coefficients for each combination of `ℓ` values and tomographic bins.
+"""
+function compute_Cℓ(w::AbstractArray{T, 3}, K::AbstractArray{T, 4}, bkgq::BackgroundQuantities, weights_χ::AbstractArray{T, 1},
+                    weights_R::AbstractArray{T, 1}, ell_prefactor::AbstractArray{T,1}) where T
+    
+    nχ = length(bkgq.χz_array)
+    Δχ = ((last(bkgq.χz_array)-first(bkgq.χz_array))/(nχ-1))
 
-end
\ No newline at end of file
+    @tullio Cℓ[l,i,j] := ell_prefactor[l]*bkgq.χz_array[n]*K[i,j,n,m]*w[l,n,m]*weights_χ[n]*weights_R[m]*Δχ
```

Comment:
I am not sure this is the best way to proceed.
In this case, you are passing the kernel (i.e. the product of the two window function) in order to compute the Cls. You might as well have passed the two windows `WA` and `WB` and written
`@tullio Cℓ[l,i,j] := ell_prefactor[l]*bkgq.χz_array[n]*WA[i,n]*WB[j,m]*w[l,n,m]*weights_χ[n]*weights_R[m]*Δχ`
The advantage is that we don't have to create and store an additional matrix, `K`. We can discuss it, if my suggestion is not ideal.